src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
         Method::Everything => None,
     };
 
-    activate_deps(cx, registry, parent, platform, deps.iter(), 0,
-                  &mut |mut cx, registry| {
-        cx.visited.remove(id);
-        finished(cx, registry)
-    })
+    Ok(Some(DepsFrame{
+        parent: parent,
+        remaining_siblings: RcVecIter::new(deps),
+        platform: platform.map(|s| s.to_string()),
+        id: id,
+    }))
 }
 
-/// Activates the dependencies for a package, one by one in turn.
-///
-/// This function will attempt to activate all possible candidates for each
-/// dependency of the package specified by `parent`. The `deps` iterator
-/// provided is an iterator over all dependencies where each element yielded
-/// informs us what the candidates are for the dependency in question.
-///
-/// The `platform` argument is the target platform that the dependencies are
-/// being activated for.
+#[derive(Clone)]
+struct RcVecIter<Elem> {
+    vec: Rc<Vec<Elem>>,
+    next_index: usize,
+}
+
+impl<Elem> RcVecIter<Elem> {
+    fn new(vec: Vec<Elem>) -> RcVecIter<Elem> {
+        RcVecIter {
+            vec: Rc::new(vec),
+            next_index: 0
+        }
+    }
+    fn next(&mut self) -> Option<(usize, &Elem)> {
+        match self.vec.get(self.next_index) {
+            None => None,
+            Some(val) => {
+                let index = self.next_index;
+                self.next_index += 1;
+                Some((index, val))
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct DepsFrame {
+    parent: Rc<Summary>,
+    remaining_siblings: RcVecIter<DepInfo>,
+    platform: Option<String>,
+    id: PackageId,
+}
+type RemainingDeps = Vec<DepsFrame>;
```

Comment:
I personally tend to prefer to avoid type aliases like this, could `Vec<DepsFrame>` be used below instead of the alias?